hail/src/main/scala/is/hail/expr/ir/TableIR.scala

```diff
@@ -1148,22 +1148,38 @@ case class TableOrderBy(child: TableIR, sortFields: IndexedSeq[SortField]) exten
   }
 }
 
-case class LocalizeEntries(child: MatrixIR, entriesFieldName: String) extends TableIR {
-  private val m = Map(MatrixType.entriesIdentifier -> entriesFieldName)
-  private val newRowType = child.typ.rvRowType.rename(m)
+/** Create a Table from a MatrixTable, storing the column values in a global
+  * field 'colsFieldName', and storing the entry values in a row field
+  * 'entriesFieldName'.
+  */
+case class CastMatrixToTable(
+  child: MatrixIR,
+  entriesFieldName: String,
+  colsFieldName: String
+) extends TableIR {
 
-  def typ: TableType = TableType(newRowType, child.typ.rowKey, child.typ.globalType)
+  def typ: TableType = TableType(
+    rowType = child.typ.rvRowType.rename(Map(MatrixType.entriesIdentifier -> entriesFieldName)),
+    key = child.typ.rowKey,
+    globalType = child.typ.globalType.appendKey(colsFieldName, +TArray(child.typ.colType)))
```

Comment:
Are we at a place where making this required won't potentially cause problems elsewhere? I think having it be required is correct since the array can't ever be missing, but if we still don't deal with requiredness well right now maybe we could hold off on that until we can?

Reply:
requiredness definitely does cause problems elsewhere. This is probably safe, but I don't think there's much of a performance reason either.

Hopefully this problem will all go away in a few weeks...